File: device/android/src/main/java/com/capacitorjs/plugins/device/DevicePlugin.kt
package com.capacitorjs.plugins.device

import android.os.Build
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.Locale

@CapacitorPlugin(name = "Device")
public class DevicePlugin : Plugin() {
    private lateinit var implementation: Device

    override fun load() {
        implementation = Device(context)
    }

    @PluginMethod
    public fun getId(call: PluginCall) {
        val r = JSObject()

        r.put("identifier", implementation.uuid)

        call.resolve(r)
    }

    @PluginMethod
    public fun getInfo(call: PluginCall) {
        val r = JSObject()

        r.put("memUsed", implementation.memUsed)
        r.put("model", Build.MODEL)
        r.put("operatingSystem", "android")
        r.put("osVersion", Build.VERSION.RELEASE)
        r.put("androidSDKVersion", Build.VERSION.SDK_INT)
        r.put("platform", implementation.platform)
        r.put("manufacturer", Build.MANUFACTURER)
        r.put("isVirtual", implementation.isVirtual)
        r.put("name", implementation.name)
        r.put("webViewVersion", implementation.webViewVersion)

        call.resolve(r)
    }

    @PluginMethod
    public fun getBatteryInfo(call: PluginCall) {
        val r = JSObject()

        // Java widened the float to a double here, which is the number JavaScript has been receiving
        r.put("batteryLevel", implementation.batteryLevel.toDouble())
        r.put("isCharging", implementation.isCharging)

        call.resolve(r)
    }

    @PluginMethod
    public fun getLanguageCode(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", Locale.getDefault().language)
        call.resolve(ret)
    }

    @PluginMethod
    public fun getLanguageTag(call: PluginCall) {
        val ret = JSObject()
        ret.put("value", Locale.getDefault().toLanguageTag())
        call.resolve(ret)
    }
}

File: device/ios/Sources/DevicePlugin/DevicePlugin.swift
import Foundation
import UIKit
import Capacitor

@objc(DevicePlugin)
public class DevicePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DevicePlugin"
    public let jsName = "Device"
    public let pluginMethods: [CAPPluginMethod] = [
        .promise("getId", DevicePlugin.getId),
        .promise("getInfo", DevicePlugin.getInfo),
        .promise("getBatteryInfo", DevicePlugin.getBatteryInfo),
        .promise("getLanguageCode", DevicePlugin.getLanguageCode),
        .promise("getLanguageTag", DevicePlugin.getLanguageTag)
    ]
    private let implementation = Device()

    func getId(_ call: CAPPluginCall) {
        if let uuid = UIDevice.current.identifierForVendor {
            call.resolve([
                "identifier": uuid.uuidString
            ])
        } else {
            call.reject("Id not available")
        }
    }
    func getInfo(_ call: CAPPluginCall) {
        var isSimulator = false
        var modelName = ""
        #if targetEnvironment(simulator)
        isSimulator = true
        modelName = ProcessInfo().environment["SIMULATOR_MODEL_IDENTIFIER"] ?? "Simulator"
        #else
        modelName = implementation.getModelName()
        #endif

        let memUsed = implementation.getMemoryUsage()
        let systemVersionNum = implementation.getSystemVersionInt() ?? 0

        call.resolve([
            "memUsed": memUsed,
            "name": UIDevice.current.name,
            "model": modelName,
            "operatingSystem": "ios",
            "osVersion": UIDevice.current.systemVersion,
            "iOSVersion": systemVersionNum,
            "platform": "ios",
            "manufacturer": "Apple",
            "isVirtual": isSimulator,
            "webViewVersion": UIDevice.current.systemVersion
        ])
    }

    func getBatteryInfo(_ call: CAPPluginCall) {
        UIDevice.current.isBatteryMonitoringEnabled = true

        call.resolve([
            "batteryLevel": UIDevice.current.batteryLevel,
            "isCharging": UIDevice.current.batteryState == .charging || UIDevice.current.batteryState == .full
        ])

        UIDevice.current.isBatteryMonitoringEnabled = false
    }

    func getLanguageCode(_ call: CAPPluginCall) {
        let code = implementation.getLanguageCode()
        call.resolve([
            "value": code
        ])
    }

    func getLanguageTag(_ call: CAPPluginCall) {
        let tag = implementation.getLanguageTag()
        call.resolve([
            "value": tag
        ])
    }

}
